src/module.c

```diff
@@ -10388,6 +10393,154 @@ int RM_GetDbIdFromDefragCtx(RedisModuleDefragCtx *ctx) {
+    if (fd < 0 || fd >= aeGetSetSize(server.el)) {
+        errno = ERANGE;
+        return REDISMODULE_ERR;
+    }
+
+    if (!callback || mask & ~(REDISMODULE_EVENTLOOP_READABLE |
+                              REDISMODULE_EVENTLOOP_WRITABLE)) {
+        errno = EINVAL;
+        return REDISMODULE_ERR;
+    }
+
+    /* We are going to register stub callbacks to 'ae' for two reasons :
+     * - "ae" callback signature is different from RedisModuleEventLoopCallback,
+     *   that will be handled it in our stub callbacks.
+     * - We need to remap 'mask' value to provide binary compatibility.
+     *
+     *  For the stub callbacks, saving user 'callback' and 'user_data' in an
+     *  EventLoopData object and passing it to ae, later we'll extract
+     *  'callback' and 'user_data' from that.
+     */
+    EventLoopData *data = aeGetFileClientData(server.el, fd);
+    if (!data)
+        data = zcalloc(sizeof(*data));
+
+    aeFileProc *aeProc;
+    if (mask & REDISMODULE_EVENTLOOP_READABLE)
+        aeProc = eventLoopCbReadable;
+    else
+        aeProc = eventLoopCbWritable;
+
+    int aeMask = eventLoopToAeMask(mask);
+
+    if (aeCreateFileEvent(server.el, fd, aeMask, aeProc, data) != AE_OK) {
+        if (aeGetFileEvents(server.el, fd) == AE_NONE)
+            zfree(data);
+        return REDISMODULE_ERR;
+    }
+
+    data->user_data = user_data;
+    if (mask & REDISMODULE_EVENTLOOP_READABLE)
+        data->rProc = callback;
+    if (mask & REDISMODULE_EVENTLOOP_WRITABLE)
+        data->wProc = callback;
+
+    errno = 0;
+    return REDISMODULE_OK;
+}
+
+/* Delete a pipe / socket event from the event loop.
+ *
+ * On success REDISMODULE_OK is returned, otherwise
+ * REDISMODULE_ERR is returned and errno is set to the following values:
+ *
+ * * ERANGE: 'fd' is negative or higher than 'maxclients' Redis config.
+ * * EINVAL: 'mask' value is invalid.
+ *            Valid values for 'mask':
+ *               REDISMODULE_EVENTLOOP_READABLE
+ *               REDISMODULE_EVENTLOOP_WRITABLE
+ *               REDISMODULE_EVENTLOOP_READABLE | REDISMODULE_EVENTLOOP_WRITABLE
+ */
+int RM_EventLoopDel(int fd, int mask) {
+    if (fd < 0 || fd >= aeGetSetSize(server.el)) {
+        errno = ERANGE;
+        return REDISMODULE_ERR;
+    }
+
+    if (mask & ~(REDISMODULE_EVENTLOOP_READABLE |
+                 REDISMODULE_EVENTLOOP_WRITABLE)) {
+        errno = EINVAL;
+        return REDISMODULE_ERR;
+    }
+
+    /* After deleting the event, if fd does not have any registered event
+     * anymore, we can free the EventLoopData object. */
+    EventLoopData *data = aeGetFileClientData(server.el, fd);
+    aeDeleteFileEvent(server.el, fd, eventLoopToAeMask(mask));
+    if (aeGetFileEvents(server.el, fd) == AE_NONE)
+        zfree(data);
+
+    errno = 0;
+    return REDISMODULE_OK;
+}
+
+/* This function can be called from other threads to wake up Redis thread */
+void RM_EventLoopWakeup(void) {
```

Comment:
What is the point of this? The event loop should keep firing as long as there is data to be read/written to.

Reply:
By using Redis event loop, module can now run fully in main thread context. Still, there might be another thread in the system, say for slow operations. Module sends operation to that thread and wants to know once the operation is completed. In that case, that thread needs to wake up event loop, so "module" can take an action. Users can do this by registering a pipe() fd to the eventloop and slow opreation thread can write a byte into it. Wakeup similar function is often required and similar function exists in popular event loop implementations, libuv, netty etc. So, I added this one as well, hoping to make eventloop api user life easier.

Reply:
Oh, I see now, its intended to trigger the serverEvent. I will say that the serverEvent is the bit I don't like very much in the API. I see other event APIs implement something more like a oneshot (thinking of tokio, but I have only played with event loops in Rust). Something like:
```
oneShot RM_CreateOneShot(OneShotCallBack cb); /* Creates something that is passed to the slow thread */
RM_FireOneShot(oneShot); /* Is called from the slow thread to trigger the callback on the main thread */
```

Maybe there is some other functionality I'm missing.